### PR TITLE
Search SDK: Adding arm-search spec for 2015-08-19 API version

### PR DESCRIPTION
**Pull request by bitsforthought**
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 
### PR information
- [X] The title of the PR is clear and informative.
- [X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [X] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [X] If applicable, the PR references the bug/issue that it fixes.
- [X] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).
### Quality of Swagger
- [X] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [X] My spec meets the review criteria:
  - [X] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [X] Validation errors from the [Linter extension for VS Code](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/linter.md) have all been fixed for this spec. (**Note:** for large, previously checked in specs, there will likely be many errors shown. Please contact our team so we can set a timeframe for fixing these errors if your PR is not going to address them).
  - [X] The spec follows the patterns described in the [Swagger good patterns](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-good-patterns.md) document unless the service API makes this impossible.

FYI @chaosrealm @seansaleh @HeidiSteen @EvanBoyle @jack4it

---

This change is [Reviewable](https://reviewable.io/reviews/azure/azure-rest-api-specs/637)

### CODE REVIEW

File: arm-search/2015-08-19/swagger/search.json
Comment (hovsepm) [inline arm-search/2015-08-19/swagger/search.json]: Listing services inside resource group will never return more than one page of results?

**Review by seansaleh — APPROVED** [review]

**Comment by brjohnstmsft** [comment]
FYI @tbombach

**Comment by brjohnstmsft** [comment]
We'll be replacing the externalDocs links to MSDN as soon as we have new content to replace it.

File: arm-search/2015-08-19/swagger/search.json
Comment (tbombach) [inline arm-search/2015-08-19/swagger/search.json]: It's nice that these constraints are described, but should they also be JSON constraints? The regex pattern might be complicated to do, but the min/max string constraints wouldn't be difficult.

File: arm-search/2015-08-19/swagger/search.json
Comment (tbombach) [inline arm-search/2015-08-19/swagger/search.json]: One note with `modelAsString: false` is if a new value is ever added, the client will be unable to handle it. When trying to deserialize the `name` property, it will be unable to map any values not listed here to a correct enum value (for languages that use real enums).

**Review by tbombach — CHANGES_REQUESTED** [review]
The JSON constraints aren't a "must-have" for merging (especially not a regex pattern that covers the rules for dashes).

The only question I have before merging is around use of enums - is it fine that the client will be unable to handle new values for that property? It would take a major version bump in the SDKs to support a new value (since enum additions are considered breaking changes in strongly typed languages). If you want to minimize breaking changes, then modeling as string will do that.

File: arm-search/2015-08-19/swagger/search.json
Comment (brjohnstmsft) [inline arm-search/2015-08-19/swagger/search.json]: Good point. One thing I don't like about modelAsString though is that it breaks Intellisense. In our data plane SDK, I created an "extensible enum" pattern to handle this, but I haven't had time to try to fold it back into AutoRest.

How about this... Since we don't plan to add any SKUs in the near future, let's leave it like this for now, and I'll use the addition of a new SKU as a forcing function add extensible enums to AutoRest. Sound good?

**Comment by brjohnstmsft** [comment]
I decided to add some JSON examples to the spec... Please hold off merging for now. Thanks!

File: arm-search/2015-08-19/swagger/search.json
Comment (tbombach) [inline arm-search/2015-08-19/swagger/search.json]: Yes, I think that's our currently only supported way of having the desired Intellisense experience for a set of known values. Let's leave it that way for now with the knowledge that a major version bump for SDKs would be required once SKU is expanded (or the extensible enums that you mentioned).

**Comment by brjohnstmsft** [comment]
@HeidiSteen I ended up adding the examples after all, since it's possible to have different examples for the request & response.

@tbombach I'm ready for this to be merged now, assuming you agree with my comments above.

File: arm-search/2015-08-19/swagger/search.json
Comment (tbombach) [inline arm-search/2015-08-19/swagger/search.json]: The benefit is that validation would fail before a call is made the service.

The constraints (including regex pattern) get turned into checks in the `Validate()` method: https://github.com/Azure/azure-sdk-for-net/blob/04d78e2101f7d040aaa97481757a2fb69fffe9c9/src/Search/Microsoft.Azure.Search/GeneratedSearchService/Models/StandardAnalyzer.cs#L65

That get called when making requests: https://github.com/Azure/azure-sdk-for-net/blob/04d78e2101f7d040aaa97481757a2fb69fffe9c9/src/Search/Microsoft.Azure.Search/GeneratedSearchService/DataSourcesOperations.cs#L102

That way a call would fail fast with a [ValidationException](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/src/ClientRuntime/Microsoft.Rest.ClientRuntime/ValidationException.cs), rather than making an actual call to the service and waiting for a response. The goal is to do client-side validation to the best of our ability, even though there's already server-side validation. It's also easier for a user writing code that uses the client - the response from the service is probably some variation of a 4xx Bad Request, while a ValidationException contains the regex pattern that failed and the target property that failed the pattern. If you rely on the response, then the customer code has to parse the message on the CloudError to find the problem.

**Comment by tbombach** [comment]
Thanks for the update @brjohnstmsft. The enum plan makes sense. I left another comment re: the regex constraint.

File: arm-search/2015-08-19/swagger/search.json
Comment (brjohnstmsft) [inline arm-search/2015-08-19/swagger/search.json]: This seems like solving the problem at the wrong layer. It introduces unnecessary coupling between the service and the client. Consider versioning -- What if we discover a bug in the regex and fix it in our RP, but forget to fix it in the client? Or maybe we fix it in both places, but some customers are still using an older client with the wrong regex. This kind of coupling seems like too high a price to pay just to shave off some latency in an error case.

Regarding usability, here is the experience of providing an invalid service name today:

```csharp
[Fact]
public void CreateServiceWithInvalidNameGivesUsefulMessage()
{
    Run(() =>
    {
        SearchManagementClient searchMgmt = GetSearchManagementClient();
        SearchService service = DefineServiceWithSku(SkuName.Free);

        CloudException e =
            Assert.Throws<CloudException>(() => searchMgmt.Services.CreateOrUpdate(Data.ResourceGroupName, InvalidServiceName, service));
        Assert.Equal("Service name '----badname' is invalid: Service name must only contain lowercase letters, digits or dashes, cannot start or end with or contain consecutive dashes and is limited to 60 characters.", e.Message);
    });
}
```

If need be, we can include the regex in the error message, but for most users this won't make much sense.

**Comment by brjohnstmsft** [comment]
@tbombach Replied to your comment above.

**Review by tbombach — APPROVED** [review]

File: arm-search/2015-08-19/swagger/search.json
Comment (seansaleh) [inline arm-search/2015-08-19/swagger/search.json]: Review opened to change it

**Pull request merged**
